Add `closed` property to `Curve3D`

GreenCrowDev opened the pull request:
Should close https://github.com/godotengine/godot-proposals/issues/6396 and close https://github.com/godotengine/godot-proposals/issues/8650

https://github.com/godotengine/godot/assets/62719360/27b8d2d2-4b51-4ea1-a221-8ea61fa51dbb

**TODO:**

- [x] Change color of start and end points (especially important when the curve is closed and you can't tell where's the start);
- [x] In and Out gizmo handles display, respectively for Start and End node, when `_closed = true`;
- [x] In and Out properties don't show up in inspector right now, but should be displayed when `_closed = true`;
- [x] ~Close Curve button could become a Closed Curve toggle, so that user can easily access the value through the already nice icon on the top bar of the viewport;~ Path3DEditorPlugin Close Curve button is now a "Toggle Open/Closed Curve" button that just toggles the curve `closed` property;
- [x] Documentation;
- [ ] Tests (can you suggest some? I'm not used to them);

**TODO (optional):**

- [ ] Draw indices text near the respective point to easily identify them (needs a bespoke solution as at the moment there's not a system in place to draw text near 3D gizmos);

If you have suggestions about what behaviour I should try to check if stuff is broken by this change let me know (for example something like `PathFollow3D`...).

GreenCrowDev committed:
Implement closed path for Curve3d

AThousandShips commented on scene/resources/curve.h:
```diff
@@ -263,6 +263,8 @@ class Curve3D : public Resource {
 	mutable Vector<size_t> points_in_cache;
 #endif
 
+	bool _closed = false;
```
```suggestion
	bool closed = false;
```
This file doesn't apply `_` prefix to member variables, let's keep it uniform,

matricola787 replied:
Oh, okay, I see. I basically copied the PR for Line2D, but I guess I misunderstood how to use these, thank you 👍

AThousandShips commented on scene/resources/curve.cpp:
```diff
@@ -2321,6 +2377,8 @@ void Curve3D::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("_get_data"), &Curve3D::_get_data);
 	ClassDB::bind_method(D_METHOD("_set_data", "data"), &Curve3D::_set_data);
 
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "_closed"), "set_closed", "is_closed");
```
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "closed"), "set_closed", "is_closed");
```
This should not be prefixed with `_` unless you intend it to be hidden in the documentation, which I don't think should be done here

matricola787 commented:
Should I commit the changes to the branch as they are ready or wait to complete all of them? I'm not sure if running CI/CD checks every time is a problem...
Right now I'm adding features to a _dev branch on top of this one.

AThousandShips commented on editor/plugins/path_3d_editor_plugin.cpp:
```diff
@@ -436,8 +443,42 @@ void Path3DGizmo::redraw() {
+			// Need to define indices separately.
+			Vector<int> idx;
+			idx.resize(pc);
+			for (int j = 0; j < pc; j++) {
+				idx.write[j] = j;
```
```suggestion
			int *idx_ptr = idx.ptrw();
			for (int j = 0; j < pc; j++) {
				idx_ptr[j] = j;
```
More efficient version, avoids COW

AThousandShips commented on editor/plugins/path_3d_editor_plugin.cpp:
```diff
@@ -436,8 +443,42 @@ void Path3DGizmo::redraw() {
+			add_handles(first_pt_handle_point, c->is_closed() ? closed_pt_handle_material : first_pt_handle_material, first_pt_id);
+
+			// Add handles for remaining intermediate points.
+			if (primary_handle_points.size()) {
```
```suggestion
			if (!primary_handle_points.is_empty()) {
```

GreenCrowDev commented:
Any idea when to merge this and set a milestone? 4.3 or 4.4?

AThousandShips commented:
It hasn't been reviewed or decided on so it's in 4.x until some decision or review has been made 🙂

Chubercik commented:
>Should I commit the changes to the branch as they are ready or wait to complete all of them? I'm not sure if running CI/CD checks every time is a problem...
>Right now I'm adding features to a _dev branch on top of this one.

No one ever reprimanded me for committing to a live PR branch, so I think it's fine - just be sure to squish your commits afterwards (see: [Pull request workflow](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html)).

Calinou reviewed (changes requested):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/29b3d9e9e538f0aa8effc8ad8bf19a2915292a89), I noticed some issues:

- The start/end color for gizmos only appears at certain camera angles. Perhaps two sprite gizmos are overlapping and only one of them is showing up?

https://github.com/godotengine/godot/assets/180032/8c311b77-6435-4136-9278-b3537c99c589

This also causes the colors to stop appearing depending on the number of points drawn:

https://github.com/godotengine/godot/assets/180032/b5a6d2a3-e152-4f8a-8c9b-53b1bb2c1cf8

They show up reliably in a top orthogonal view, which is what leads me to think it's an overlapping gizmo sorting issue:

![image](https://github.com/godotengine/godot/assets/180032/3ded2244-fcd3-4cf7-b67f-c2a2ee530a2d)

Also, as an improvement, I suggest drawing the last line that goes from the last point to the first in a different color:

![image](https://github.com/godotengine/godot/assets/180032/67c5eaff-795b-452f-a03a-850c4816b830)

Mockup made in GIMP:

![image](https://github.com/godotengine/godot/assets/180032/57e2fe51-1a02-453f-93e2-f2d9675702f2)

This way, you can tell which line is automatically being inferred by the engine as opposed as being designed by the user.

GreenCrowDev commented:
Thanks for the review @Calinou!

Yes, this worked well before https://github.com/godotengine/godot/pull/87901, and there's this other PR https://github.com/godotengine/godot/pull/89847 that refines `Path3D` editor, ready to be merged.

Since they work on the same lines of code, I'd like to wait for https://github.com/godotengine/godot/pull/89847 to be merged, that defines stuff the lack of which makes difficult to finalize these last tweaks you mentioned.

Is that okay? 😊

Chubercik commented:
> Since they work on the same lines of code, I'd like to wait for #89847 to be merged, that defines stuff the lack of which makes difficult to finalize these last tweaks you mentioned.

![Let's go.](https://github.com/godotengine/godot/assets/37378746/ca56a177-440c-4f4f-a999-5c97c1e11e0e)

GreenCrowDev commented:
I fixed the handles issue @Calinou.

Also fixed handles color in the curve edit mode, after the rebase.

I didn't implement the last segment color change though, as I realized that it would require some decent work, since we don't have access to control points but the baked ones. Eventually to add in a later PR? The bones that describe the curve direction should be sufficient, in the meantime.

GreenCrowDev commented:
I fixed the handles issue @Calinou.

Also fixed handles color in the curve edit mode, after the rebase.

I didn't implement the last segment color change though, as I realized that it would require some decent work, since we don't have access to control points but the baked ones. Eventually to add in a later PR? The bones that describe the curve direction should be sufficient, in the meantime.

AThousandShips commented on scene/resources/curve.cpp:
```diff
@@ -2274,13 +2340,13 @@ void Curve3D::_get_property_list(List<PropertyInfo> *p_list) const {
 		p_list->push_back(pi);
 
-		if (i != 0) {
+		if (i != 0 || is_closed()) {
```
```suggestion
		if (closed || i != 0) {
```

AThousandShips commented on scene/resources/curve.cpp:
```diff
@@ -2274,13 +2340,13 @@ void Curve3D::_get_property_list(List<PropertyInfo> *p_list) const {
 		}
 
-		if (i != points.size() - 1) {
+		if (i != points.size() - 1 || is_closed()) {
```
```suggestion
		if (closed || i != points.size() - 1) {
```

GreenCrowDev commented:
Changes implemented.

Also fixed a couple of bugs.

Not sure why it fails the check on Linux 🤔
```
gpg: keyserver receive failed: Connection reset by peer
Error: Process completed with exit code 2.
```

AyyZerrAsa commented:
> I fixed the handles issue @Calinou.
> 
> Also fixed handles color in the curve edit mode, after the rebase.
> 
> I didn't implement the last segment color change though, as I realized that it would require some decent work, since we don't have access to control points but the baked ones. Eventually to add in a later PR?

What do you mean in regards to not having access to control points? I made some changes in #90357 that caches baked distance at each original control point. Not sure whether that's relevant here. Great commit btw I was needing this, super suprised closed curves weren't a thing already.

Calinou reviewed (approved):
Tested locally (rebased on top of `master` 79de2eaae7fb3bb8f13031ac2ec9b19128b6f471), it works as expected. I think this is in a good state to be merged early in the 4.4 cycle.

https://github.com/godotengine/godot/assets/180032/2888f45a-0b1b-4999-a437-cdcbfa566f02

Calinou commented on editor/plugins/path_3d_editor_plugin.cpp:
```diff
@@ -281,8 +281,15 @@ void Path3DGizmo::redraw() {
+	Ref<StandardMaterial3D> closed_pt_handle_material = gizmo_plugin->get_material("closed_pt_handle");
 	Ref<StandardMaterial3D> sec_handles_material = gizmo_plugin->get_material("sec_handles");
 
+	first_pt_handle_material->set_albedo(Color(0.0, 1.0, 0.0));
+	last_pt_handle_material->set_albedo(Color(1.0, 0.0, 0.0));
+	closed_pt_handle_material->set_albedo(Color(1.0, 1.0, 0.0));
```
I find the pure yellow a bit hard to distinguish from white, maybe try a slightly orange color for the closed handle:

```suggestion
	first_pt_handle_material->set_albedo(Color(0.2, 1.0, 0.0));
	last_pt_handle_material->set_albedo(Color(1.0, 0.2, 0.0));
	closed_pt_handle_material->set_albedo(Color(1.0, 0.8, 0.0));
```

This also includes tweaks to the first/last point colors to look more visually pleasing.

![image](https://github.com/godotengine/godot/assets/180032/4a62cea4-fbe9-45aa-acd9-d1cfcf6f1ea5)

![image](https://github.com/godotengine/godot/assets/180032/b9571b69-b174-4a80-b1e8-a9b6bf9fa52a)

Calinou commented on editor/plugins/path_3d_editor_plugin.cpp:
```diff
@@ -934,6 +973,14 @@ void Path3DGizmoPlugin::redraw(EditorNode3DGizmo *p_gizmo) {
+	Ref<StandardMaterial3D> last_pt_handle_material = get_material("last_pt_handle", p_gizmo);
+	Ref<StandardMaterial3D> closed_pt_handle_material = get_material("closed_pt_handle", p_gizmo);
+
+	first_pt_handle_material->set_albedo(Color(0.0, 1.0, 0.0));
+	last_pt_handle_material->set_albedo(Color(1.0, 0.0, 0.0));
+	closed_pt_handle_material->set_albedo(Color(1.0, 1.0, 0.0));
```
Same as above:

```suggestion
	first_pt_handle_material->set_albedo(Color(0.2, 1.0, 0.0));
	last_pt_handle_material->set_albedo(Color(1.0, 0.2, 0.0));
	closed_pt_handle_material->set_albedo(Color(1.0, 0.8, 0.0));
```

GreenCrowDev commented:
Thanks for the review!
I applied the suggestions and rebased :)

GreenCrowDev commented:
Rebased :)

akien-mga commented:
This seems fine overall to me, but I question the usefulness of the `toggle_closed` method, which is equivalent to `closed = !closed`. Is this used often enough to use a dedicated method? I note that Line2D has a `closed` property but no `toggle_closed` method.

GreenCrowDev commented:
> This seems fine overall to me, but I question the usefulness of the `toggle_closed` method, which is equivalent to `closed = !closed`. Is this used often enough to use a dedicated method? I note that Line2D has a `closed` property but no `toggle_closed` method.

I honestly don't recall why `toggle_closed` is there, I guess I'll remove it and force push again.

AThousandShips commented on scene/resources/curve.cpp:
```diff
@@ -1596,13 +1609,16 @@ void Curve3D::_bake() const {
 		points_in_cache.set(0, 0);
 #endif
 
+		// Point Count: Begins at 1 to account for the last point
```
```suggestion
		// Point Count: Begins at 1 to account for the last point.
```

Repiteo commented:
Thanks!
